src/Razor/src/Microsoft.AspNetCore.Razor.Language/DefaultRazorTagHelperBinderPhase.cs

```diff
@@ -63,6 +77,37 @@ protected override void ExecuteCore(RazorCodeDocument codeDocument)
             codeDocument.SetSyntaxTree(rewrittenSyntaxTree);
         }
 
+        private string ComputeNamespace(RazorCodeDocument codeDocument)
+        {
+            var rootNamespace = codeDocument.GetCodeGenerationOptions().RootNamespace;
+            var relativePath = codeDocument.Source.RelativePath;
+            if (string.IsNullOrEmpty(rootNamespace) || string.IsNullOrEmpty(relativePath))
+            {
+                return null;
+            }
+
+            // Sanitize the base namespace, but leave the dots.
+            var builder = new StringBuilder();
+            var segments = rootNamespace.Split(NamespaceSeparators, StringSplitOptions.RemoveEmptyEntries);
+            builder.Append(CSharpIdentifier.SanitizeIdentifier(segments[0]));
+            for (var i = 1; i < segments.Length; i++)
+            {
+                builder.Append('.');
+                builder.Append(CSharpIdentifier.SanitizeIdentifier(segments[i]));
+            }
+
+            segments = relativePath.Split(PathSeparators, StringSplitOptions.RemoveEmptyEntries);
+
+            // Skip the last segment because it's the FileName.
+            for (var i = 0; i < segments.Length - 1; i++)
+            {
+                builder.Append('.');
+                builder.Append(CSharpIdentifier.SanitizeIdentifier(segments[i]));
+            }
+
+            return builder.ToString();
+        }
```

Comment:
Yeah I'd really like to see this pull out into a static class with some unit tests. We're going to need this functionality forever.